# Making Recommendations Based on Popularity

These datasets are hosted on: https://archive.ics.uci.edu/ml/datasets/Restaurant+%26+consumer+data

They were originally published by: Blanca Vargas-Govea, Juan Gabriel González-Serna, Rafael Ponce-Medellín. Effects of relevant contextual features in the performance of a restaurant recommender system. In RecSys11: Workshop on Context Aware Recommender Systems (CARS-2011), Chicago, IL, USA, October 23, 2011.

## Restaurants data

In [1]:
import numpy as np
import pandas as pd

In [2]:
# rating_final.csv
url = 'https://drive.google.com/file/d/1ptu4AlEXO4qQ8GytxKHoeuS1y4l_zWkC/view?usp=sharing' 
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
frame = pd.read_csv(path)

# chefmozcuisine.csv
url = 'https://drive.google.com/file/d/1S0_EGSRERIkSKW4D8xHPGZMqvlhuUzp1/view?usp=sharing' 
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
cuisine = pd.read_csv(path)

# 'geoplaces2.csv'
url = 'https://drive.google.com/file/d/1ee3ib7LqGsMUksY68SD9yBItRvTFELxo/view?usp=sharing' 
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
geodata = pd.read_csv(path, encoding = 'CP1252') # change encoding to 'mbcs' in Windows

On the `frame` dataset we have the ratings users have given to places. Ratings go from 0 to 2.

In [3]:
frame.head(3)

userID  placeID  rating  food_rating  service_rating
0  U1077   135085       2            2               2
1  U1077   135038       2            2               1
2  U1077   132825       2            2               2

In the `geodata` dataset we have info about the places. We will only use the `name` column.

In [4]:
geodata.head(2)

placeID   latitude   longitude  \
0   134999  18.915421  -99.184871   
1   132825  22.147392 -100.983092   

                                      the_geom_meter             name  \
0  0101000020957F000088568DE356715AC138C0A525FC46...  Kiku Cuernavaca   
1  0101000020957F00001AD016568C4858C1243261274BA5...  puesto de tacos   

                                  address        city    state country fax  \
0                              Revolucion  Cuernavaca  Morelos  Mexico   ?   
1  esquina santos degollado y leon guzman      s.l.p.   s.l.p.  mexico   ?   

   ...            alcohol smoking_area dress_code     accessibility   price  \
0  ...  No_Alcohol_Served         none   informal  no_accessibility  medium   
1  ...  No_Alcohol_Served         none   informal        completely     low   

                     url Rambience franchise    area other_services  
0  kikucuernavaca.com.mx  familiar         f  closed           none  
1                      ?  familiar         f    open           none  

[2 rows x 21 columns]

In [5]:
#we are gonna use only place ID and name
places =  geodata[['placeID', 'name']]
places.head()

placeID                            name
0   134999                 Kiku Cuernavaca
1   132825                 puesto de tacos
2   135106      El Rincón de San Francisco
3   132667  little pizza Emilio Portes Gil
4   132613                   carnitas_mata

In the `cuisine` dataset we have the type of cuisine that restaurants offer.

In [6]:
cuisine.head(3)

placeID        Rcuisine
0   135110         Spanish
1   135109         Italian
2   135107  Latin_American

## Popularity/Quality based recommmender system

Let's group places by rating, and look at their average rating. This is an **explicit** rating given by users.

In [7]:
rating = pd.DataFrame(frame.groupby('placeID')['rating'].mean())
rating.sort_values("rating", ascending=False).head()

rating
placeID          
132955   2.000000
135034   2.000000
134986   2.000000
132922   1.833333
132755   1.800000

The top rated places have a perfect score of 2/2. But how many reviews do these places have?

In [8]:
frame.query("placeID==132955")

userID  placeID  rating  food_rating  service_rating
934   U1004   132955       2            2               2
960   U1061   132955       2            2               2
996   U1059   132955       2            1               2
1014  U1097   132955       2            2               1
1080  U1096   132955       2            2               2

Looks like only 5 people went to this place. Maybe they're just the owner's friends! Or maybe they're really top-quality places, but too niche to recommend to the masses.

We can also look at how many times each restaurant has received a rating. The ratings count is an **implicit** rating.

In [9]:
rating['rating_count'] = frame.groupby('placeID')['rating'].count()
rating.sort_values("rating_count", ascending=False).head()

rating  rating_count
placeID                        
135085   1.333333            36
132825   1.281250            32
135032   1.178571            28
135052   1.280000            25
132834   1.000000            25

Some places have been visited around 30 times. They are more popular than the top rated places, but received lower explicit ratings.

Let's locate the most popular place, and get some info about it:

In [10]:
rating.sort_values('rating_count', ascending=False).head()

rating  rating_count
placeID                        
135085   1.333333            36
132825   1.281250            32
135032   1.178571            28
135052   1.280000            25
132834   1.000000            25

In [11]:
# placeId of most popular place
top_popular_placeID = rating.sort_values('rating_count', ascending=False).head(1).index[0]

# name of the most popular place
places[places['placeID']==top_popular_placeID]

placeID                    name
121   135085  Tortas Locas Hipocampo

In [12]:
# cuisine of the most popular place
cuisine[cuisine['placeID']==top_popular_placeID]

placeID   Rcuisine
44   135085  Fast_Food

The most popular place is "Tortas Locas Hipocampo", a fast food place that has received 36 reviews and it has an average score of 1.33.

### Challenge:

Find a hybrid system to sort restaurants, so that you can recommend the "best" places: restaurants that are both high rated and popular.

* Weighted average
$$ w_i = \frac{ c_i \cdot r_i}{\sum_i c_i} $$
where $w_i$ is the new hybrid measure, $c_i$ and $r_i$ the counts and rating of the $i$-th system.

* Linear combination: we assign different weight to counts and ratings and then sum

$$ \ell_i = a c_i + b r_i$$

In [13]:
rating.head(2)

rating  rating_count
placeID                      
132560     0.50             4
132561     0.75             4

In [14]:
def weight_hybrid(n,df):
    
    #this function adds a new column with the weights and returns the "heaviest" n resturants
    
    df2 = df.copy() 
    df2['weight'] = (df['rating_count'] * df['rating']) / (df['rating_count'].sum())
    
    return df2.sort_values(by="weight", ascending = False).head(n)

weight_hybrid(10,rating)

rating  rating_count    weight
placeID                                  
135085   1.333333            36  0.041344
132825   1.281250            32  0.035314
135032   1.178571            28  0.028424
135052   1.280000            25  0.027562
135038   1.208333            24  0.024978
135062   1.238095            21  0.022394
135042   1.250000            20  0.021533
132862   1.388889            18  0.021533
132834   1.000000            25  0.021533
135060   1.136364            22  0.021533

In [16]:
def linear_hybrid(n, df, weight_counts):
    # This function linearly combines ratings and counts with appropriate weights
        
    if weight_counts < 0 or weight_counts > 1:
        print("Weight must be in [0, 1]")
        
    col_name = f"lin. {weight_counts*100}%"
    df2 = df.copy() 
    df2[col_name] = weight_counts * df['rating_count'] + (1 - weight_counts) * df['rating']
    
    return df2.sort_values(by=col_name, ascending=False).head(n)

linear_hybrid(10,rating, 0.7)

rating  rating_count  lin. 70.0%
placeID                                    
135085   1.333333            36   25.600000
132825   1.281250            32   22.784375
135032   1.178571            28   19.953571
135052   1.280000            25   17.884000
132834   1.000000            25   17.800000
135038   1.208333            24   17.162500
135060   1.136364            22   15.740909
135062   1.238095            21   15.071429
135042   1.250000            20   14.375000
132862   1.388889            18   13.016667